File: go/05-functions/methods.go
package main

import "fmt"

// Rectangle is a simple struct with a width and a height.
type Rectangle struct {
    Width  float64
    Height float64
}

// Area is a method with a value receiver; it reads but does not modify.
func (r Rectangle) Area() float64 {
    return r.Width * r.Height
}

// Scale is a method with a pointer receiver; it modifies the struct.
func (r *Rectangle) Scale(factor float64) {
    r.Width *= factor
    r.Height *= factor
}

func main() {
    rect := Rectangle{Width: 3, Height: 4}
    fmt.Printf("Area: %.1f\n", rect.Area())

    rect.Scale(2)
    fmt.Printf("After scaling: %.1f x %.1f\n", rect.Width, rect.Height)
    fmt.Printf("New area: %.1f\n", rect.Area())
}

File: kotlin/05-functions/Methods.kt
// Member functions (methods) and extension functions

class Rectangle(val width: Int, val height: Int) {
    // A method — belongs to the class and can use its properties
    fun area(): Int = width * height

    // Methods can call other methods of the same class
    fun describe(): String = "Rectangle ${width}x$height with area ${area()}"
}

// An extension function adds a method to the String type
fun String.shout(): String = this.uppercase() + "!"

// An extension function on Int
fun Int.isEven(): Boolean = this % 2 == 0

fun main() {
    val rect = Rectangle(4, 5)
    println(rect.area())        // call a method with dot notation
    println(rect.describe())

    // Extension functions are called exactly like built-in methods
    println("hello".shout())
    println("10 is even: ${10.isEven()}")
    println("7 is even: ${7.isEven()}")
}
